app/roster/page.tsx

```diff
@@ -0,0 +1,128 @@
+"use server";
+
+import { auth } from "@clerk/nextjs/server";
+
+import { getRosterCmsData } from "@/lib/server/cms-data";
+import { Hero, Level, type Artifact } from "@/lib/roster";
+import Artifacts from "@/components/roster/artifacts/Artifacts";
+import Levels from "@/components/roster/Levels";
+import {
+  getRosterArtifacts,
+  getRosterHeroes,
+  getRosterLevels,
+} from "@/lib/server/roster";
+import Heroes from "@/components/roster/heroes/Heroes";
+import { getUser } from "@/lib/server/users";
+import UserAvatar from "@/components/UserAvatar";
+import { AscensionLevel } from "@/lib/characters";
+import { Card, CardContent } from "@/components/ui/card";
+import { Tabs, TabsList, TabsTrigger, TabsContent } from "@/components/ui/tabs";
+
+export default async function MyRoster() {
+  const { userId } = auth();
+
+  if (!userId) {
+    auth().redirectToSignIn();
+    return;
+  }
+
+  const { user_image, username } = await getUser(userId);
+
+  const rosterJsonData = await getRosterCmsData();
+
+  const rosterLevels = await getRosterLevels(userId);
+  const rosterHeroes = await getRosterHeroes(userId);
+  const rosterArtifacts = await getRosterArtifacts(userId);
+
+  function getLevelForLevelsFromData(key: string): number {
+    let level = rosterLevels.find(
+      (level) => level.levelId.toString() === key,
+    )?.level;
+
+    return level ? level : 0;
+  }
+
+  const levels: Level[] = Object.entries(rosterJsonData.seasons).map(
+    ([key, data]: [string, any]) => ({
+      key: key,
+      seasonName: data["name"],
+      maxLevel: data["maxLevel"],
+      active: data["active"],
+      level: getLevelForLevelsFromData(key),
```

Comment:
maybe i'm not understanding the data model correctly, but i'm wondering why this lookup is necessary?

Reply:
So this, and the same for the artifacts, I build the objects for each part of the roster. This "lookup" is for the part that's user specific to populate what's in the database if it exists. Does this make sense? Keep in mind it's doing a lookup on rosterArtifacts, rosterLevels, and rosterHeroes. I guess I could have just done it like I did the heroes? Don't need a separate function per se, but "separation of concerns" right? Technically the lookup has nothing to do with building the list, it could do without it.

Reply:
okay this seems fine to me